sw/host/opentitanlib/src/rom_patch/patch.rs

```diff
@@ -0,0 +1,745 @@
+        let raw_header = self as *const _ as *const u8;
+        let header_size = mem::size_of::<RomPatchHeader>();
+
+        if buf.len() < header_size {
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        unsafe {
+            ptr::copy_nonoverlapping(raw_header, buf.as_mut_ptr(), header_size);
+        }
+
+        Ok(header_size)
+    }
+}
+
+impl fmt::Display for RomPatchHeader {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let unaligned = std::ptr::addr_of!(self.size);
+        let size = unsafe { std::ptr::read_unaligned(unaligned) };
+        write!(
+            f,
+            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
+            self.meta, self.revision, size
+        )
+    }
+}
+
+struct RomPatchSubRoutine {
+    base_addr: u32,
+    offset: usize,
+    code: Vec<u32>,
+}
+
+impl RomPatchSubRoutine {
+    fn from_hjson(routine: &RomPatchSubRoutineHjson, patch_base_addr: u32) -> Result<Self> {
+        let base_addr = u32::from_str(&routine.base_addr)?;
+        if base_addr < patch_base_addr {
+            bail!(
+                "Invalid subroutine base addr {:#x} < {:#x} ",
+                base_addr,
+                patch_base_addr
+            );
+        }
+
+        let mut code = vec![];
+        for insn in &routine.code {
+            let insn_hex = u32::from_str(insn)?;
+            code.push(insn_hex);
+        }
+
+        let offset = PATCH_CODE_SECTION_OFFSET + (base_addr - patch_base_addr) as usize;
+
+        Ok(Self {
+            base_addr,
+            offset,
+            code,
+        })
+    }
+
+    fn len(&self) -> usize {
+        self.code.len() * 4
+    }
+
+    fn offset(&self) -> usize {
+        self.offset
+    }
+
+    fn range(&self) -> Range<u32> {
+        Range {
+            start: self.base_addr,
+            end: self.base_addr + (self.code.len() as u32 * 4),
+        }
+    }
+}
+
+impl fmt::Display for RomPatchSubRoutine {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let range = self.range();
+        write!(
+            f,
+            "\tBase address {:#04x} Range [{:#04x}:{:#04x}]",
+            self.base_addr, range.start, range.end
+        )
+    }
+}
+
+impl Read for RomPatchSubRoutine {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.len() < self.len() {
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        let mut buf_offset = 0;
+
+        for insn in &self.code {
+            let insn_bytes = insn.to_le_bytes();
+            let raw_insn_bytes = &insn_bytes as *const _ as *const u8;
+            unsafe {
+                ptr::copy_nonoverlapping(raw_insn_bytes, buf[buf_offset..].as_mut_ptr(), 4);
+            }
+            buf_offset += 4;
+        }
+
+        Ok(self.len())
+    }
+}
+
+struct RomPatch<'a> {
+    header: RomPatchHeader,
+    size: usize,
```

Comment:
maybe document what this is the size of? And size in what unit (bytes?)

Reply:
Removed from the updated PR.